### PR TITLE
Add a graph mapper to delete Derived directory for each of the projects of the graph

### PR DESCRIPTION
**Pull request by pepibumur**
### Short description 📝

This PR adds a new graph mapper, `DeleteDerivedDirectoryGraphMapper`, that returns side effects to delete the derived directory for each project of the graph. By deleting the directories first, we prevent that we end up with leftover files from previous runs. 

I'm aware we this is not the most efficient solution since we could diff the new derived files, compare them with the existing ones, and get the IO operations that need to be executed. Indeed, this is the path that I'm taking with this. Once we have all the IO operations modeled as side effects, we'll be able to group them, and determine how they translate to IO operations based on what's existing in the system. For instance, the following 2 operations:

```
delete directory at /project/Derived
create file at /project/Derived/InfoPlists/App.plist
```
Can be merged into just:

```
create file at /project/Derived/InfoPlists/App.plist
```

Note that the mapper is defined in the `TuistGenerator` framework sot that projects depending on the framework can use it too.

### CODE REVIEW

File: Sources/TuistCore/Descriptors/CommandDescriptor.swift [hunk 1]
```diff
@@ -6,12 +6,24 @@ import Foundation
...
 
     /// Creates a command descriptor
     /// - Parameter command: The command and its arguments to perform
     public init(command: [String]) {
         self.command = command
     }
+
+    // MARK: - Hashable
+
+    public func hash(into hasher: inout Hasher) {
```
Comment (kwridan) [inline Sources/TuistCore/Descriptors/CommandDescriptor.swift:20]: there isn't a need to manually write this out - the compiler should be able to synthesize this automatically :)

File: Sources/TuistCore/Graph/Mappers/GraphMapper.swift [hunk 2]
```diff
@@ -5,21 +5,21 @@ import Foundation
 public protocol GraphMapping {
     /// Given a graph, it maps it into another graph.
     /// - Parameter graph: Graph to be mapped.
-    func map(graph: Graph) throws -> (Graph, [SideEffectDescriptor])
+    func map(graph: Graph) throws -> (Graph, Set<SideEffectDescriptor>)
```
Comment (kwridan) [inline Sources/TuistCore/Graph/Mappers/GraphMapper.swift:8]: The order in which side effects are performed matters

e.g.

- delete derived directory
- create derived directory
- create file

Placing those in a set will not guarantee the correct execution of steps :/

File: Sources/TuistGenerator/Generator/DerivedFileGenerator.swift [hunk 3]
```diff
@@ -14,12 +14,11 @@ protocol DerivedFileGenerating {
     /// - Throws: An error if the generation of the derived files errors.
     /// - Returns: A project that might have got mutated after the generation of derived files, and a
     ///     function to be called after the project generation to delete the derived files that are not necessary anymore.
-    func generate(graph: Graph, project: Project, sourceRootPath: AbsolutePath) throws -> (Project, [SideEffectDescriptor])
+    func generate(graph: Graph, project: Project, sourceRootPath: AbsolutePath) throws -> (Project, Set<SideEffectDescriptor>)
```
Comment (kwridan) [inline Sources/TuistGenerator/Generator/DerivedFileGenerator.swift:17]: for a separate PR - we can try modeling this as a `GeneratedInfoPlistGraphMapper` - it was left in the generator to ease the migration to descriptors and we didn't have the mapper concept ironed out quite yet then :) 

That should help simplify this (and avoid it attempting to mutate the project mid generation!)

File: Sources/TuistGenerator/GraphMappers/DeleteDerivedDirectoryGraphMapper.swift [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+public class DeleteDerivedDirectoryGraphMapper: GraphMapping {
+    public init() {}
+
+    // MARK: - GraphMapping
+
+    public func map(graph: Graph) throws -> (Graph, Set<SideEffectDescriptor>) {
+        logger.debug("Determining the /Derived directories that should be delted")
+        var sideEffects = Set<SideEffectDescriptor>()
+
+        graph.projects.forEach { project in
```
Comment (kwridan) [inline Sources/TuistGenerator/GraphMappers/DeleteDerivedDirectoryGraphMapper.swift:15]: this whole mappers concept is turning out to be very beautiful!

File: Sources/TuistCore/Descriptors/CommandDescriptor.swift (hunk 1 above)
Comment (kwridan) [reply Sources/TuistCore/Descriptors/CommandDescriptor.swift:20]: (the same with the other `hash(into` methods

File: Sources/TuistCore/Graph/Mappers/GraphMapper.swift (hunk 2 above)
Comment (pepibumur) [reply Sources/TuistCore/Graph/Mappers/GraphMapper.swift:8]: Ouch, I missed that one 🤦

File: Sources/TuistGenerator/GraphMappers/DeleteDerivedDirectoryGraphMapper.swift (hunk 4 above)
Comment (pepibumur) [reply Sources/TuistGenerator/GraphMappers/DeleteDerivedDirectoryGraphMapper.swift:15]: It is. I love it a lot ❤️

**Comment by pepibumur** [comment]
I addressed your comments here @kwridan

**Comment by pepibumur** [comment]
@kwridan in fact, I had to implement `GenerateInfoPlistGraphMapper` and make sure it runs right after deleting the Derived directory. Could you give it another look? 👁️

File: Sources/TuistCore/Graph/Mappers/TargetNodeGraphMapper.swift [hunk 5]
```diff
@@ -8,73 +8,79 @@ import Foundation
 /// any transformations will no longer be part of the resulting graph.
 ///
 /// - Note: When mapping, the `transform` block receives a copy of the original `TargetNode`
-public class TargetNodeGraphMapper: GraphMapping {
-    public let mapTargetNode: (TargetNode) -> TargetNode
-    public init(transform: @escaping (TargetNode) -> TargetNode) {
+open class TargetNodeGraphMapper: GraphMapping {
```
[the remark this answers is not in the document]
Comment (pepibumur) [reply Sources/TuistCore/Graph/Mappers/TargetNodeGraphMapper.swift:11]: Not really. I might have made it open while trying different approaches. I reverted it.

**Comment by pepibumur** [comment]
@kwridan made `Project` a struct and fixed the logic to replace the target. Let's see if CI is happy now :)

**Comment by kwridan** [comment]
Ouch, I think we've hit the limitation of `TargetNodeGraphMapper` it can't modify projects as a single project can be referenced by several `TargetNode`s 😞.

I had a play to try out "model mappers" - pushed a draft PR to show case it #1227 something like that can help model this operation a bit better and potentially some of the others we've encountered recently like the ones we do in in the `enrich` method here: 

https://github.com/tuist/tuist/blob/master/Sources/TuistLoader/Loaders/GeneratorModelLoader.swift#L52

Let me know what you think, if it looks good can split it out into smaller reviewable chunks.

**Comment by kwridan** [comment]
@pepibumur I've pushed #1469 as a continuation of this based on model mappers

**Pull request closed without merging**